src/pmacfiltercontrol/hdfadapter.py

```diff
@@ -0,0 +1,106 @@
+import h5py
+import os
+from numpy import int64 as np_int64
+
+from typing import Optional
+
+ATTENUATION_KEY = "attenuation"
+ADJUSTMENT_KEY = "adjustment"
+FRAME_NUMBER_KEY = "frame_number"
+UID_KEY = "/uid"
+
+
+class HDFAdapter:
+    def __init__(
+        self,
+        hdf_file_path: str = "",
+    ) -> None:
+
+        self.file_path: str = hdf_file_path
+        self.file: Optional[h5py.File] = None
+
+    def _set_file_path(self, new_file_path: str) -> None:
+
+        if self._check_path(new_file_path):
+            self.file_path = new_file_path
+
+    def _open_file(self) -> None:
+        if self.file is None:
+            if self._check_path():
+                self.file = h5py.File(self.file_path, "w", libver="latest")
+                print(f"* File {self.file} is open.")
+                self._setup_datasets()
+        else:
+            if self.file_path != self.file.filename:
+                print("* Another file is already open and being written to.")
+
+    def _close_file(self) -> None:
+        if self.file is not None:
+            try:
+                assert isinstance(self.file, h5py.File)
+                print(f"* File {self.file} has been closed.")
+                self.file.close()
+                self.file = None
+            except Exception as e:
+                print(f"* Failed closing file.\n{e}")
+        else:
+            print(f"* No file is open, ignoring close...")
+
+    def _check_path(self, file_path: str) -> bool:
+        if file_path == "" or file_path is None:
+            print(f"* Please enter a valid file path.\nPath={self.file_path}")
+        else:
+            parent_path: str = file_path.rsplit("/", 1)[0]
+            if not os.path.isdir(parent_path):
+                print("* Path not found. Enter a valid path.")
+            else:
+                return True
+
+        return False
+
+    def _setup_datasets(self) -> None:
+
+        print(f"* Creating/fetching datasets in HDF5 file: {self.file_path}")
+
+        def _fetch_dataset(key: str) -> h5py.Dataset:
+            dset: h5py.Dataset = None
+            if key not in self.file.keys():
+                dset = self.file.create_dataset(key, (1,), maxshape=(None,), dtype=int)
+            else:
+                dset = self.file.get(key)
+            return dset
+
+        self.adjustment_dset = _fetch_dataset(ADJUSTMENT_KEY)
+        self.attenuation_dset = _fetch_dataset(ATTENUATION_KEY)
+        self.uid_dataset = _fetch_dataset(UID_KEY)
```

Comment:
I think this method should just create the 3 datasets and then once self.file is not None you can assume the datasets are created.

Reply:
This function is only called when the file is first opened, and it checks if the datasets already exist (i.e. there's already data to append to). It's more a safety feature so data isn't overwritten accidentally or errors are raised as it tries to make a dataset that already exists.

Reply:
> it checks if the datasets already exist (i.e. there's already data to append to)

I think it would be better to not allow opening an existing file at all, which it should never do anyway.

> with just checking if self.file is not None, it could be open, but not have any datasets set up for some reason, so I can't just assume they exist without checking

But if you create the file and create the datasets in an atomic way (i.e. in one method) so that it is not possible, then you can make that assumption and not have to do the checks.